File: Path Sum/Path Sum.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a boolean
    def hasPathSum(self, root, sum):
    	return self.dfs(root, sum, 0)


    def dfs(self, node, sum, current_sum):
    	if node == None:
    		return False;
    	if node.left == None and node.right == None:
    		return current_sum + node.val == sum
    	return self.dfs(node.left, sum, current_sum + node.val) or self.dfs(node.right, sum, current_sum + node.val)

File: Path Sum/Path Sum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode *root, int sum) {
        return dfs(root, sum, 0);
    }
    
    bool dfs(TreeNode *node, int sum, int currentSum){
        if (NULL == node){
            return false;
        }
        bool result1;
        bool result2;
        if (node->left == NULL && node->right == NULL){
            return sum == currentSum + node->val;
        }
        if (node->left != NULL){
            result1 = dfs(node->left, sum, currentSum + node->val);
        }
        if (node->right != NULL){
            result2 = dfs(node->right, sum, currentSum + node->val);
        }
        return result1 || result2;
        // return dfs(node->left, sum, currentSum + node->val) || dfs(node->right, sum, currentSum + node->val);
    }
};